File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {

        int maxi=0;
        
        int n=prices.size();
        int prev=prices[n-1];
        for(int i=n-1;i>=0;i--){
            int profit=prev-prices[i];
            maxi=max(maxi,profit);
            if(profit<0){
                prev=prices[i];
            }
        }
        return maxi;

    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int prev = prices[n-1];
        int maxi =0;

        for(int i=n-1;i>=0;i--){
           
            int profit = prev - prices[i];
          
            maxi = maxi < profit ? profit : maxi;
            if(profit<0) prev = prices[i];

        }
        return maxi;
    }
}